doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,243 @@
+      `timer_reset()` like in Contiki is fine)
+    - Long term low resolution software timers that play well with power management
+    - Short term high resolution software timers
+- Delay of program execution
+    - API to block calling thread for long periods and allows other threads to run
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable, hardware independent
+      units
+        - At least: seconds, milli seconds
+    - UNIX epoch as common reference point for high interoperability
+    - Support for 64 bit time stamps
+        - Devices running for detects might want to log events with milli second
+          resolution timestamps on flash / sd card / ...
+        - Note: The rest of the API can be 32 bit, only for time stamps 64 bit
+          support is mandatory
+    - Optional feature to auto-initialize system time from RTC/RTT during boot
+    - APIs to allow implementing network time synchronization
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+    - System time utilities
+        - Functions to convert between different time formats, e.g. UNIX and GPS timestamps
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time. More
+  specifically: With *n* software timers used in parallel, the longest period of
+  time with interrupts disabled be at most *O(n)*
+- The implementation must aim for a minimized number of IRQs caused
+- The implementation must aim for keeping its ISRs short; with *n* software
+  timers used in parallel, the ISR should run for at most *O(n)*
+- The implementation should provide means to execute callbacks in thread
+  context
+
+## Precision Requirements
+
+- Timer callbacks should never fire prematurely (±1 tick of the underlying
+  timer)
+- The accuracy of execution delay should be measured with single digit micro
+  seconds for short delays in absence of other load on the system
+    - It is acceptable if platform specific tweaking is needed to achieve this
+    - It is acceptable if minimum hardware requirements are needed to achieve
+      this
+- Timer callbacks should fire closely after the timer expired
+    - In absence of other load on the system, accuracy in the order of single
+      digit micro seconds should be achievable when given hardware requirements
+      are fulfilled and platform specific tweaking is done
+
+## Performance Requirements
+
+- Efficient use of RAM/ROM
+- Optimized for power management
+    - The timer implementation should minimize system wake ups
```

Comment:
Most MCUs have two kind of timers

 - coarse, low-power timers
 - high precision timers with higher power requirement

Now a timer API could combine both to always set the longest possible sleep with the low power timer and use the high precision timer for the rest.

e.g. if you have a low power 32kHz timer and a fast 1 MHz timer and want to sleep for 325 ticks and then use the high precision timer for the remaining 82 ticks.

How ideally you could also specify a minimum and a maximum wake up time, so wake-up events can be bundled to reduce the amount of total wake-ups.

For precise timings you'd have `min == max`

Reply:
> e.g. if you have a low power 32kHz timer and a fast 1 MHz timer and want to sleep for 325 ticks and then use the high precision timer for the remaining 82 ticks.

I'm not sure if there really is a use case for that. E.g. when you want to perform 3 measurements with a delay of exactly 10 ms and additionally want to repeat those three measurements every 30 minutes, this would be a good example for using both timers. But to me it makes more sense to have one 30 minutes timer, and then a 10 ms timer that is used for the second and third measurement of each row.

If you would transparently use both low frequency and high frequency timers (e.g. for a sleep of 30 minutes and 1337 µs, transparently set one 30 minute low frequency timer and afterwards a 1337 µs high frequency timer), there would be some issues: 

1. The overhead of setting the high frequency timer would reduce precision
2. Low and high frequency domain might be driven from different crystals, so there might be a clock drift between those two. (In fact, having a separate 32.678 kHz crystal for the low frequency timer is quite common.)
3. The chance to start a timer right on the clock tick is pretty low. In the worst case are off by just bellow clock tick. For a 1 MHz timer that would be up to 1 µs precision loss. On a 32.678 kHz timer that would be ~30 µs precision loss.

IMO mixing clocks should not be done transparently, as this might (depending on the availability of low frequency clocks) reduce the accuracy of timers in unexpected ways. Having this explicitly makes sure users are aware of power consumption, accuracy and precision of the used clock.

> How ideally you could also specify a minimum and a maximum wake up time, so wake-up events can be bundled to reduce the amount of total wake-ups.

Sounds like this adds quite some complexity. I'm not sure if it is better to let the application developers optimize their use of timers. IMO core infrastructure like software timers should come with low ROM / RAM requirements and robust implementations, as one usually cannot opt out of using them.

Reply:
Yeah, this is something that seems useful in theory. In practice, in a non-idle system, it is difficult enough to get a 1MHz timer actually have better than the 30.5us accuracy that a 32kHz timer promises. See https://github.com/RIOT-OS/RIOT/pull/14931. We tried debugging this, uncorrected, the (quite fast) nrf52840 has 30us offset on a simple ztimer_sleep(). There's a context switch, expensive timer reads, and ztimer's overhead. The time is spread more or less evenly across them. Yeah, that can be mostly corrected, but this is the simplest use-case.

A simple combination of timers won't gain anything, as setting a low-freq timer already quantizes on that timer's frequency, so the result is off by half that timer's period length, on average, anyhow. So using a HF timer "for the last microseconds", but being based on a, from the HF timer's perspective, arbitrary time in between a LF period, is useless.
There are ways to double sync, but that's expensive...

IMO, with all these theoretical timing needs, people need to realize that in most cases, 1KHz or 32786KHz is mightily sufficient. And those are easy to do with RTTs.

Reply:
Also, as said in previous discussions, a sane abstraction of timers that use only a single backend (e.g., don't do these complex combinations of timers) are a nice base to put the complex stuff __on top__. At least for prototyping.

Reply:
> But the user shouldn't have to worry about which timer to use, that should be selected automatically based on the wait time.

The user needs to choose the precision (the time base). There cannot be an API where the user says "sleep 100". It'll always be "sleep 100ms" or "sleep 100us".

IMO it is OK to simplify our lives (and reduce complexity) by asking the user to always choose the time base that is just enough to express the needs. E.g., if the user asks for "sleep 100ms", it is acceptable to expect them to call "sleep_ms(100)", and not do "sleep_us(100000)".

Reply:
Adding another opinion: the arguments @benpicco raised and the underlying ideas should (and can) not be dismissed like that.

> 1. The overhead of setting the high frequency timer would reduce precision

This overhead should be well defined. If this overhead is not tolerable it is also not tolerable for using just the HF timer.

> 2 Low and high frequency domain might be driven from different crystals, so there might be a clock drift between those two. (In fact, having a separate 32.678 kHz crystal for the low frequency timer is quite common.)

True. But not relevant for deciding if splitting across timers (domains) makes sense:
1) You'd know how the timeout is split between both domains, thus drift boundaries of both can be combined to give a respective boundary for the compound.
2) In fact, using the more accurate LF crystal for the significant (long) part of the timeout *improves* the overall drift (versus perfect reference) compared to only using a worse accuracy HF clock.
3) Clock drifts can be determined and compensated or calibrated

This: ..
> The chance to start a timer right on the clock tick is pretty low. In the worst case are off by just bellow clock tick. For a 1 MHz timer that would be up to 1 µs precision loss. On a 32.678 kHz timer that would be ~30 µs precision loss.

.. and this ..

> A simple combination of timers won't gain anything, as setting a low-freq timer already quantizes on that timer's frequency, so the result is off by half that timer's period length, on average, anyhow. So using a HF timer "for the last microseconds", but being based on a, from the HF timer's perspective, arbitrary time in between a LF period, is useless.
There are ways to double sync, but that's expensive...

... seem to be only related to "thinking inside of the box" ;)
Your quantization problem is only there because you assume there is no way to relate the absolute time references of the LF timer to the HF timer. If you know when exactly a timeout of n ticks on the LF timer will fire (i.e. where in the period it is) you can adjust the remaining ticks for the HF timer accordingly.

> IMO, with all these theoretical timing needs, people need to realize that in most cases, 1KHz or 32786KHz is mightily sufficient. And those are easy to do with RTTs.

Agree! Only: "most cases" != "all cases other developers may rightfully care about"
We should not decide what makes sense for others. If something is not possible, well it's not possible. But if it **is** possible and reasonable we should not rule out or restrict it because we have different priorities.

> (..) nice base to put the complex stuff on top

Fully agree as long as the base doesn't limit us to still do the possible & reasonable things others deem more important than the things we consider as "most cases".

> (..) it is acceptable to expect them to call "sleep_ms(100)", and not do "sleep_us(100000)".

Kind of. Still, regarding power consumption one might argue that it would be easier to write *save as much energy as possible*-applications in a platform independent way if the abstraction does whatever is best on the current platform. The restrictions that made you decide xx ms is the best threshold to always map to timer A instead of timer B may not be the same on another platform and the desired timeout may be variable. I thought these kind of considerations are of interest to go for "the RIOT way": make it easy for the developer AND provide flexibility if you want to adapt things more towards something you explicitly need. Maybe my interpretation is off here?

Sometimes writing can be interpreted in many ways, so:  all just my opinion, I don't have negative feelings towards any of you, I'm open minded, we probably all want the best solution :v: 
Are you guys available and interested in some joint discussion of these things during the summit? I don't think we need a separate breakout session for this but maybe chatting about this during a coffe-break or something helps understanding the different perspectives better? What do you think?

Reply:
I think this is re-hashing the mailing list discussions from a year ago :)

I'd be happy to see a prototype of what is proposed here.
I agree that for "sleep_ms(100)" vs "sleep_us(100000)", there are ways to maximize power efficiency. A simple `if us > 1000: sleep_ms(us/1000), sleep_us(rest)` might do it for many cases.

Otherwise, I see split between domains, clock drift boundaries, high precision LF, mentioned. Implement! :)

> ... seem to be only related to "thinking inside of the box" ;)
> If you know when exactly a timeout of n ticks on the LF timer will fire (i.e. where in the period it is) you can adjust the remaining ticks for the HF timer accordingly.

Well, how do you do that without explicit synchronization? Assuming LF is being used as it is active during sleep, and HF is not active during sleep, they need to be synchronized *at least* on every transition between sleep and wake. I call that double sync.

I call that expensive because on some platforms (e.g., sam0), just reading the RTT "now" register needs to synchronize the clocks, in some cases taking __a whole period__ (30us). __sometimes__, depending on what else is going on in the system. That dwarfs a lot of clock drift for reasonable time scales.

> In fact, using the more accurate LF crystal for the significant (long) part of the timeout improves the overall drift (versus perfect reference) compared to only using a worse accuracy HF clock.

Yup. I'm strongly in favor of not making things complex, and teach users to use ms or 32kHz, unless there's a reeeeally good reason. Accurate ms or 32kHz timing goes very far. :)

Reply:
> Are you guys available and interested in some joint discussion of these things during the summit? I don't think we need a separate breakout session for this but maybe chatting about this during a coffe-break or something helps understanding the different perspectives better? What do you think?

Sure!

Reply:
> I think this is re-hashing the mailing list discussions from a year ago :)

I know, and to be honest I don't want to bore people with lengthy discussions again. There is no point in these lengthy "abstract" discussions if not more of the developers jump in.  Yet, the arguments are not less valid than before. It appears to me that we all hold single pieces of the big picture without an explicit agreement on how the big picture looks like. This document will eventually help with that I think.

> Otherwise, I see split between domains, clock drift boundaries, high precision LF, mentioned. Implement! :)

On it (at least for some of these problems) just from a slightly different direction.. I'll explain during the summit.

> Well, how do you do that without explicit synchronization? (..)

Not. At least not without any synchronization. Of course they need to be synchronized.
But why double sync? And who says that you need to do this explicitly? We may very well use an "opportunistic" approach where you sync based on events that happen anyway. I.e., you don't care how long reading 'now' takes for the slow timer if you read the fast timer close enough to a known event of the slow timer... I don't say this solves all cases everywhere. I'm just trying to think towards a solution instead of searching for reasons against it.

> I call that expensive

Yes reading LF timer can indeed be expensive. Apart from reading not being necessarily required, some would argue transitioning to a low power mode for arbitrarily small periods is "expensive" because of the delay when leaving lpm. Again, it all depends on the perspective ;)

> I'm strongly in favor of not making things complex, and teach users to use ms (..)

Fair enough. I get your point.

Reply:
BTW: why do I get these check-labels mail-notification after adding a comment? 
> [RIOT-OS/RIOT] Run failed: check-labels - timer-rdm (f5fae3e)

Reply:
In the end this will have to be a trade-off between:

1. Overhead in terms of CPU cycles
2. RAM and ROM consumption
3. Complexity and maintainability of the implementation
4. Give the developer using the API a solid understanding and full control about accuracy, resolution, and impact on power consumption for a timer
5. Simplify API so that `sleep_ms(1)` is just an alias for `sleep_us(1000)`
6. Allow optimization of power consumption, accuracy, etc. without replacing code. (E.g. via `USEMODULE += ztimer_policy_powersafe` sleeps could be moved to the low frequency clock where possible, etc. And one would not have to replace `ztimer_sleep(ZTIMER_USEC, x)` by `ztimer_sleep(ZTIMER_MSEC, (x + 500) / 1000)`)

We just won't be able to get those all.

IMO we should first assemble a list design goals and vote on them to get some agreement of a prioritization of them. If overhead, memory, and maintainability get a higher priority than user friendliness and timer policy support, an explicit clock selection via the API makes more sense. Otherwise, adding heuristics and policies for clock selection makes more sense. (This might still be technically better implemented on top of a "low-level high-level timer API", but that would be an implementation detail.)